File: solutions/808/main.py
from functools import lru_cache

class Solution:
    def soupServings(self, n: int) -> float:
        if n > 5000: 
            return 1.0 

        # NOTE, need to set the maxsize large 
        @lru_cache(maxsize=1000)
        def dfs(a, b): 
            # base case 
            if a == 0 and b != 0: 
                return 1.0 
            if a == 0 and b == 0: 
                return 0.5 
            if a != 0 and b == 0: 
                return 0 

            subtract = lambda x, y: x-min(x, y)

            prob = 0 
            prob += 0.25 * dfs(subtract(a, 100), b)
            prob += 0.25 * dfs(subtract(a, 75), subtract(b, 25))
            prob += 0.25 * dfs(subtract(a, 50), subtract(b, 50))
            prob += 0.25 * dfs(subtract(a, 25), subtract(b, 75))

            return prob 

        return dfs(n, n)

if __name__ == "__main__": 
    s = Solution() 

    assert s.soupServings(50) == 0.62500 
    assert s.soupServings(100) == 0.71875

File: solutions/808/main.rs
struct Solution {}

impl Solution {
    pub fn dfs(memo: &mut Vec<Vec<Option<f64>>>, a: i32, b: i32, opt: &impl Fn(i32, i32)->i32) -> f64 {
        // base case 
        if a == 0 && b != 0 {
            return 1.0; 
        }
        if a == 0 && b == 0 {
            return 0.5; 
        }
        if a != 0 && b == 0 {
            return 0.0; 
        }

        match memo[a as usize][b as usize] {
            Some(prob) => {
                return prob; 
            }
            None => {}
        }

        let mut prob = 0.0; 
        for i in 0..=3 {
            let amount = (i as i32) * 25;
            prob += 0.25 * Solution::dfs(memo, opt(a, 100-amount), opt(b, amount), opt);
        }
        memo[a as usize][b as usize] = Some(prob); 
        
        return prob; 
    }

    pub fn soup_servings(n: i32) -> f64 {
        if n > 5000 {
            return 1.0; 
        }

        let mut memo = vec![vec![None; n as usize+1]; n as usize+1]; 

        let subtract = |x: i32, y: i32| {
            return x - x.min(y); 
        };

        Solution::dfs(&mut memo, n, n, &subtract) 
    }
}

fn main() {
    assert_eq!(Solution::soup_servings(50), 0.62500); 
}